### ISSUE TITLE
Location field in CDR object can lead to missing EVSE information

### ISSUE THREAD

**Issue by fedragon** [ocpi/ocpi#149]
`CDR` object's `location` field is defined as a `Location` object of cardinality `1` but the `EVSE` field defined in it has a cardinality of `0 ... n`, which means that a CPO might:
- send an empty list of EVSEs, or
- send multiple EVSEs

I think that both situations are not ideal in case of CDRs: given that a CDR is meant for billing purposes, it is important (and necessary) that **one and exactly one** `EVSE` (and related `Connector`) is provided: the one on which the session happened. 
This is stated in the `location` field description (`Location where the charging session took place, including only the relevant EVSE and Connector`), but it is not enforced in any way by types so it can easily lead to diverging implementations.

I would propose to define a dedicated object that contains (at least) the following subset of `Location` fields:

### CDRLocation object

|Property | Type | Cardinality|
| --- | --- | --- | 
| id  | string(39) | 1 |
| address | string(45) | 1 |
| city | string (45) | 1 |
| postal_code | string(10) | 1 |
| country | string(3) | 1 |
| evse | EVSE | 1 |
| (additional fields, if relevant) | |

**Comment by RobertdeLeeuw-IHomer** [comment]
Yes, we are indeed not enforcing this via the object definition. The idea was to send the full object, and limit it by what is written in the specification. 

Are you experiencing problems in the field? 
Companies that read the specification differently?

If there are problems, I see 3 different possible solutions here:
1: We improve the wording to make it even explicit that you have to include exactly 1 EVSE and Connector, nothing more or nothing less.
This does not enfore it via the object difinition, but if everybody reads the specification the same way it works.

2: We introduce new objects as you propose above. Question then is what fields to we need. But it will work, it is just a breaking change compared to 2.1.1

3: We only provide the Location, EVSE and Connector IDs. Drawback: MSP also need to implement the Locations module.

**Comment by fedragon** [comment]
I'm not experiencing problems right now (I have yet to implement the CDR module) but I have recently had to deal with another specification being implemented in slightly different ways by different companies because it relied on requirements written in comments rather than enforced via types, thus leading to mistakes / interpretations. I understand that new objects would "pollute" the namespace but I strongly believe that they would also make the protocol more robust (and more easily testable by computers).

**Comment by mbayings** [comment]
@RobertdeLeeuw-IHomer Solution 3 is not the most desired one I think, as there is already a request to reduce the forced links to different modules. Solution 1 is easiest and can be done in simple documentation update. Solution 2 can only be done in next version.

**Comment by simonschllng** [comment]
**Upvote for option 1!**

And every party who cares about should implement a quick sanity check on the POST CDR method and reject everything with not exactly one EVSE and Connector entry. Having at least some companies implementing that check will help to get everyone do it the right way. Together with a proper description in the specs.

**Comment by fedragon** [comment]
I have a few questions about Solution 1, I hope somebody can answer them: 
- what is the advantage of allowing a CPO to POST a CDR that cannot actually be used by an eMSP? (besides the fact there is no need to roll out a new version)
- how would an eMSP communicate to a CPO that their CDR did not pass the "quick sanity check"? 
- how would/should the CPO react to a rejected session?

My point is that, by avoiding a breaking change, we introduce additional alternative flows that would not exist if types would not allow for ambiguity. Leaving the responsibility of implementing "quick sanity checks" to users of the protocol is, in my humble opinion, no more than avoiding to pay technical debt now only to have it hit us back when the specification gets adopted.

**Comment by RobertdeLeeuw-IHomer** [comment]
We do not allow a CDR without an EVSE etc. We require it via the description in the document. We just do not enforce it via the object model.

A MSP needs to do a sanity check anyway. You can put in a lot of data that makes no sense at all. What would an MSP do when it receives a CDR of 5 minutes with 5 Megawatthour of energy used...

Yes we do need to add a way for a MSP to reject a CDR. We just were not able to find a solution before we finished 2.1, that is why this is on the list for 2.2. I had hoped we would have started the 2.2 developement last year...
For now we agreed that it would be the seem as with the Excelsheets that are now used to communicate CDRs. This will be a manual action for now. When you find something wrong a person contacts the CPO.

**Comment by mbayings** [comment]
This is the whole discussion Robert. You should not be allowed to reject a CDR. If you don't like the CDR you must request a credit-CDR and a new CDR. You cannot reject or modify it, as it is used for billing. So we must add a kind of CDR credit function.

**Comment by simonschllng** [comment]
> how would an eMSP communicate to a CPO that their CDR did not pass the "quick sanity check"?
With an appropriate HTTP error code, as you would do when anything else is wrong.

There is a way to "reject" and "accept" CDRs in OCHP, which (unfortunately) is rarely implemented. I hear from MSPs that the CDR quality is often very low and that they have to validate and solve issues manually _because_ of this missing functionality. (In general, not only OCPI)

From my pov what @RobertdeLeeuw-IHomer is saying is just it:
> Yes we do need to add a way for a MSP to reject a CDR.

**Comment by mbayings** [comment]
The bad quality of CDRs is indeed an issue. Rubish in = Rubish out. This is not solved by rejecting CDRs. In fact if CDRs cannot be rejected, but when a credit-CDR need to be made by the CPO, the CPO will probably do a better check next time when they sent it. 
Besides that when we go more and more to direct billing, a bad CDR would result in bad billing to the EV Driver. So the problem must be solved and that is that CPOs should be enforced to sent good CDRs. 
Can you imagine this in the Telco of Fuel card business?? Where the MSP must solve the rubish and problems from CPO's?

**Comment by poohsen** [comment]
I'm voting for solution 1 in the short term and solution 2 in OCPI 2.2. It also goes well with what I wrote in #153 which is about a similar issue

**Comment by fedragon** [comment]
I fully agree with @mbayings. By the way, the "credit CDR" feature he is proposing sounds to me a lot like the natural MSP counterpart to the "correction" mechanism described in #55
